[AssemblyScript] porting source from JS to AS

StEvUgnIn opened the pull request:
Follows #420

Commit:
configuring env for working with assemblyscript

Commit:
porting js into as

Commit:
set assemblyscript portability for AS/TS

Commit:
update gitignore - yarn

Commit:
configuration for compilation of AS into JS

StEvUgnIn commented:
Next steps are: 
- [ ] Fix typings between JS and AS/TS
- [ ] Port JS spec to AS/TS spec (spec.ts)
- [ ] Be able to work fully with emitted JS from AS implementation
- [ ] Write an encapsulated interface layer for using WASM from any web programming environment
- [ ] Benchmark performances between JS and AS modes

Please comment here if you are interested to generously help.

Commit:
add source types declarations for build

Commit:
fix required explicit toString conversion

Commit:
fix module typings

Commit:
implement workaround for AS100

Commit:
tsc build dist/index.dts from portable AS directly

Commit:
build successfully javascript and webassembly

trusktr commented:
This will be neat for AS users to have!

Is there a particular reason to commit the build artefacts to the repo? Those can be built during publish time when publishing to NPM for example.

Why build Wasm modules? 

If people will import the JS code into their JS projects, or import the AS code into their AS projects, then the module builds would not be needed. The Wasm modules are needed only for JS users who would import the Wasm modules instead of the JS code. 

Wondering if that's it's worth building modules. Would the barrier between JS and Wasm allow this to be faster?

It would be sweet to see what perf diff looks like in a JS example that makes many calls into the module, vs plain JS and vs plain AS.

StEvUgnIn commented:
Hi @trusktr 

> This will be neat for AS users to have!
> 

Thanks! I hope we could see more WebGL frameworks emerge on the WebAssembly platform and this helps to train programming in Assemblyscript 🙂

> Is there a particular reason to commit the build artefacts to the repo? Those can be built during publish time when publishing to NPM for example.

I was unsure at first, but I decided to include this because of the AssemblyScript conventions and I suppose someone could just download the file to include it in an existing project, especially during this testing phase. 

> 
> Why build Wasm modules?
> 
> If people will import the JS code into their JS projects, or import the AS code into their AS projects, then the module builds would not be needed. The Wasm modules are needed only for JS users who would import the Wasm modules instead of the JS code.
> 

Yes, exactly.

A default `.gitignore` lies in the `build` repository which actually lets you keep the built modules by default.

> Wondering if that's it's worth building modules. Would the barrier between JS and Wasm allow this to be faster?
> 
> It would be sweet to see what perf diff looks like in a JS example that makes many calls into the module, vs plain JS and vs plain AS.

I am looking forward to try both implementations from the browser 😃

Commit:
removed directive for TS typings

Commit:
another fix to AS100 for expected closure

Commit:
fix AS warnings

Commit:
optimize MathUtil library

MaxGraey commented on asconfig.json:
```diff
@@ -0,0 +1,20 @@
+      "textFile": "build/optimized.wat",
+      "sourceMap": true,
+      "optimizeLevel": 3,
+      "shrinkLevel": 1,
+      "converge": false,
+      "noAssert": false
```
```suggestion
      "shrinkLevel": 0,
      "converge": false,
      "noAssert": true
```
if you required best perf

StEvUgnIn replied:
Thanks! @MaxGraey

Commit:
optimize build for speed

- disable size optimization
- disable assertions in production

Co-authored-by: Max Graey <maxgraey@gmail.com>

Commit:
fix modules in dist/index.d.ts

StEvUgnIn commented:
I have a weird compiler output due to type declared inside modules yet..
https://gist.github.com/StEvUgnIn/acbfed089a24b1c671f79b1a36723a06

But so far, it seems to go well.

Commit:
compile from single file index.ts

Commit:
changetype is not implemented for TS compilation

StEvUgnIn commented:
![image](https://user-images.githubusercontent.com/7686916/113019046-77b39a80-9181-11eb-9210-89cf3cf2cc43.png)
JavaScript output now renders

Commit:
change MathUtil into JSMath and Maths namespaces

Commit:
merge JSMath and Maths and toolchain TS into AMD modules

production build achieved from assembly

Commit:
simplify toolchain to compile TS/AS into JS

Commit:
definitive toolchain compilation

esm and cjs are produced by tsc
rollup build all wasm loaders

StEvUgnIn reviewed (commented):
This work finalizes the first build targeting Javascript and Webassembly.

Commit:
hypot cannot be parsed in JS build

must declare hypot3, hypot4, hypot7, ... each for function overloading
moved namespace Math into maths.ts

StEvUgnIn reviewed (commented):
![msedge_BdF7kMQpT4](https://user-images.githubusercontent.com/7686916/113412453-9d76b480-93b8-11eb-936b-1e5a3de99bb1.png)

Commit:
javascript interface for webassembly

rollup supports operation for both release and debug
script allows to produce the exports for each loader files
loader is compiled from JS using typescript but babel is supported

Commit:
update .travis.yml for using latest node-lts version

StEvUgnIn commented on loader/index.js:
```diff
@@ -0,0 +1,14 @@
+
+let modules;
+
+wasm({ ...imports }).then(instance => {
```
This is temporary, before I implement a better solution for exporting WebAssembly modules. Solutions are either:

- Complying on @assemblyscript/loader using `loader.instantiateStreaming`
- Making a WASM file for each module (curently max size for synchronous compilation is 4KB). Current WASM file is >100KB with `-O3` optimization
- Develop a processor script to pass `<script>` to WebAssembly using `eval` and `<noscript>`

StEvUgnIn replied:
### [WebAssembly.Module.exports()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module)

StEvUgnIn replied:
I'm currently experimenting with [as-bind](https://github.com/torch2424/as-bind)

Commit:
add upper case support

Commit:
right asynchronous call for WebAssembly loading

Commit:
update assembly and asconfig.json

global function min cannot be accessed inside class when overriden
asconfig.json: tdsFiles, exportRuntime, runtime set to incremental

Commit:
loader

moved loader release and debug to assembly/loader
use as-bind for importing AS wasm
loader produces declaration files only

Commit:
build

prepare build system for as-pect files

Commit:
as-pect

default configuration
copied and renamed mocha tests from *.js into *.spec.ts

mrchantey commented:
## 32 Bit Support
This PR is great! I'm in the process of implementing the brach in my project now. 
I'd just like to flag the 64 bit operations. We may want to have flexibility, for either allowing users to make the call at runtime, or exporting two seperate builds.
For starters we could declare something like the following in `common.ts` and then use it throughout the library.
```ts
//64 
export type Float = f64
export type FloatArray = Float64Array
export const FloatMath = Math

//32
export type Float = f32
export type FloatArray = Float32Array
export const FloatMath = Mathf
```

StEvUgnIn commented:
Hello @mrchantey,

Please you are welcome for trying and sharing your proposition.

StEvUgnIn commented:
I didn't make any advancement on this PR. I was waiting for `as-bind` to support Array/Float32Array/Float64Array before going any further.

psnet commented:
I tested on another project same idea for calculations on WASM and due to the bridge JS - WASM where non-scalar values are copied between memory of JS and WASM so it is actually slower than done in pure JS.

So this entire idea is not efficient

StEvUgnIn commented:
@psnet Thank you for your feedback. I stopped working on this port a few years back already, but I am keeping an eye on ongoing discussions.
